apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/ValueComparerTest.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright 2018, Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.Firestore.V1Beta1;
+using System.Collections.Generic;
+using Xunit;
+
+namespace Google.Cloud.Firestore.Tests
+{
+    using System.Linq;
+    using static ProtoHelpers;
+
+    public class ValueComparerTest
+    {
+        // Taken from OrderTest.java, with a few extra tests.
+        // These are compared pairwise.
+        private static readonly dynamic[] s_valuesInOrder =
+        {
+            // Null
+            CreateNullValue(),
+
+            // Booleans
+            false,
+            true,
+
+            // Numbers
+            double.NaN,
+            double.NegativeInfinity,
+            double.MinValue,
+            long.MinValue,
+            int.MinValue,
+            -1.1,
+            -1L,
+            0L,
+            double.Epsilon,
+            0.1,
+            1L,
+            1.1,
+            int.MaxValue,
```

Comment:
A test value of something like `(double)(long.MaxValue - 1)` should be added here to check for integer/double comparison when the integer will lose precision when converted to a double.